### ISSUE TITLE
Steam DXVK games can't be run on AMD Vulkan Driver from Proton 5.13-2

### ISSUE THREAD

**Issue by Kai-AMD** [ValveSoftware/steam-runtime#313]
# Compatibility Report
- Name of the game with compatibility issues: Wolfenstein II, Witcher3, Strange Brigade and ...
- Steam AppID of the game: 612880, 292030, 312670 and ...

## System Information
- GPU: All AMD GPUs Supported
- Driver: AMDGPU-PRO 20.45 or open source AMDVLK v-2020.Q4.5 (Not Mesa/Radv)
- Kernel version: All Linux Kernel Supported
- Full system information: Ubuntu 20.04 + Kernel 5.4.0-53-generic
- Proton version: 5.13-2

## I confirm:
- [*] that I haven't found an existing compatibility report for this game.
- [*] that I have checked whether there are updates for my system available.

## Symptoms 
DXVK games crash with 5.13-2, but pass with 5.0-10.

Crash info:
(1) Something regarding libGL error
(2) Something regarding VK_ERROR_EXTENSION_NOT_PRESENT error

The reason maybe OpenGL/Vulkan libraries (Loader and ICD) in /opt can NOT be accessed in game:

/opt/amdgpu-pro/lib/x86_64-linux-gnu/libGL.so.1.2
/opt/amdgpu-pro/lib/x86_64-linux-gnu/libGL.so -> libGL.so.1.2
/opt/amdgpu-pro/lib/x86_64-linux-gnu/libGL.so.1 -> libGL.so.1.2
/usr/lib/x86_64-linux-gnu/dri/amdgpu_dri.so
/opt/amdgpu/lib/x86_64-linux-gnu/dri/amdgpu_dri.so -> /usr/lib/x86_64-linux-gnu/dri/amdgpu_dri.so

/usr/lib/x86_64-linux-gnu/libvulkan.so.1.2.148
/usr/lib/x86_64-linux-gnu/libvulkan.so -> libvulkan.so.1
/usr/lib/x86_64-linux-gnu/libvulkan.so.1 -> libvulkan.so.1.2.148
/opt/amdgpu-pro/lib/x86_64-linux-gnu/amdvlk64.so

Issue can be workarounded by following steps:
(1) Add “PROTON_DUMP_DEBUG_COMMANDS=1 %command%” in the launch options
(2) Run game to generate a file called "run" at /tmp/proton_root
(3) Run with command “/tmp/proton_root/run”

Issue can also be workarounded by:
(1) Change the "library_path" in json file of AMD Vulkan ICD from "/opt/amdgpu-pro/lib/x86_64-linux-gnu/amdvlk64.so" to "/usr/lib/x86_64-linux-gnu/amdvlk64.so".
(2) Copy amdvlk64.so from /opt/amdgpu-pro/lib/x86_64-linux-gnu/ to /usr/lib/x86_64-linux-gnu/, then launch steam games.

## Reproduction
(1) Install Ubuntu 20.04 64 bit OS
(2) Install AMDGPU-PRO 20.45 or set up open source AMDVLK v-2020.Q4.5
(3) Run DXVK steam games and see the crash

**Comment by smcv** [comment]
Please show us a log of what pressure-vessel is thinking, and exactly what happens. You can do this without involving Proton like this:

```
cd /path/to/SteamLinuxRuntime_soldier
PRESSURE_VESSEL_VERBOSE=1 ./run -- steam-runtime-system-info --verbose 2>&1 | tee container.log
```

and then send container.log as a gist. You can edit/censor the log if there's anything in it that you consider private, as long as it's obvious where it has been edited, for instance replacing your username with REDACTED.

The SteamLinuxRuntime_soldier directory will be in one of your Steam libraries. The most likely place is ~/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier if you haven't reconfigured the installation path.

**Comment by smcv** [comment]
Also please show us the system information (Help -> System Information in Steam). Again, you can edit/censor the log if you need to, and please send it as a [gist](https://gist.github.com/).

**Comment by WenqingLiAMD** [comment]
Hi @smcv, please get the container.log and system info here: 
https://gist.github.com/WenqingLiAMD/3736fad777675675cb85b0217557f271

**Comment by smcv** [comment]
> Hi @smcv, please get the container.log and system info here:
https://gist.github.com/WenqingLiAMD/3736fad777675675cb85b0217557f271

OK. Outside the container, the amdvlk ICD is configured like this:

```
  "vulkan" : {
    "icds" : [
      {
        "json_path" : "/usr/share/vulkan/icd.d/amd_icd32.json",
        "library_path" : "/opt/amdgpu-pro/lib/i386-linux-gnu/amdvlk32.so",
        "api_version" : "1.2.139"
      },
      {
        "json_path" : "/usr/share/vulkan/icd.d/amd_icd64.json",
        "library_path" : "/opt/amdgpu-pro/lib/x86_64-linux-gnu/amdvlk64.so",
        "api_version" : "1.2.157"
      }
    ],
```

So we would expect each word-size to successfully load its own driver, and fail to load the other word-size's driver (64-bit games will fail to load `amdvlk32.so` but successfully load `amdvlk64.so`). This is similar to how Mesa works.

Inside the container, we rewrite the JSON to have different paths:

```
  "vulkan" : {
    "icds" : [
      {
        "json_path" : "/overrides/share/vulkan/icd.d/0-i386-linux-gnu.json",
        "library_path" : "/overrides/lib/i386-linux-gnu/vulkan/0/amdvlk32.so",
        "api_version" : "1.2.139"
      },
      {
        "json_path" : "/overrides/share/vulkan/icd.d/1-x86_64-linux-gnu.json",
        "library_path" : "/overrides/lib/x86_64-linux-gnu/vulkan/1/amdvlk64.so",
        "api_version" : "1.2.157"
      }
    ],
```

where `/overrides/lib/i386-linux-gnu/vulkan/0/amdvlk32.so` is a symbolic link to `/opt/amdgpu-pro/lib/i386-linux-gnu/amdvlk32.so`, and `/overrides/lib/x86_64-linux-gnu/vulkan/1/amdvlk64.so` is a symbolic link to `/opt/amdgpu-pro/lib/x86_64-linux-gnu/amdvlk64.so`.

The missing piece (I think) is that we should be mounting the actual drivers into `/opt` in the container, and we are not.

We might also need to mount `/opt/amdgpu/share/libdrm/amdgpu.ids`.

**Comment by smcv** [comment]
If you run Steam with the environment variable `PRESSURE_VESSEL_FILESYSTEMS_RO=/opt/amdgpu:/opt/amdgpu-pro` set, does that successfully work around this issue?

Note that it is not enough to put this variable in the Launch Options, because the container is created before the launch options take effect. You have to run something like `PRESSURE_VESSEL_FILESYSTEMS_RO=/opt/amdgpu:/opt/amdgpu-pro steam`.

Whether this works or not, but especially if it doesn't, a log captured like this

```
cd /path/to/SteamLinuxRuntime_soldier
PRESSURE_VESSEL_FILESYSTEMS_RO=/opt/amdgpu:/opt/amdgpu-pro \
PRESSURE_VESSEL_VERBOSE=1 \
./run -- steam-runtime-system-info --verbose 2>&1 | tee with-opt.log
```

would be useful.

**Comment by smcv** [comment]
If that workaround is successful, then I think https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/192 should fix this.

**Comment by WenqingLiAMD** [comment]
Yes, the workaround works, log updated here: https://gist.github.com/WenqingLiAMD/3736fad777675675cb85b0217557f271#file-with-opt-log

**Comment by smcv** [comment]
Thanks! Use the workaround for now, but the next pressure-vessel release (next SteamLinuxRuntime_soldier beta) should make it unnecessary.

**Comment by Kai-AMD** [comment]
Hi @smcv, Thanks for fixing it. Could you please let me know which exact version of SteamLinuxRuntime_soldier will fix this issue? We will verify the fix :)

**Comment by smcv** [comment]
No, I can't, because it doesn't exist yet :-) When it exists, I'll comment here.

**Comment by smcv** [comment]
Today's `SteamLinuxRuntime_soldier` beta should fix this, without needing your workaround. Look for pressure-vessel 0.20210105.0 in VERSIONS.txt.

**Comment by Kai-AMD** [comment]
Thanks for fixing it! We will verify it.

**Comment by Kai-AMD** [comment]
Hi @smcv,
Could you please check if the files in /etc/amd are accessible with the beta version?
Thanks,
Kai

**Comment by smcv** [comment]
> Could you please check if the files in /etc/amd are accessible with the beta version?

They are not. Do they need to be?

In general, everything in the container's `/etc` and `/usr` comes from the container, except for files that are *specifically* taken from the host.

**Comment by Kai-AMD** [comment]
> > Could you please check if the files in /etc/amd are accessible with the beta version?
> 
> They are not. Do they need to be?
> 
> In general, everything in the container's `/etc` and `/usr` comes from the container, except for files that are _specifically_ taken from the host.

Yes, the AMD driver needs to access /etc/amd.

**Comment by smcv** [comment]
Please try creating a symbolic link `SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd`, and check whether the driver can read files from `/etc/amd` that way.

If that's successful, then we can set up future versions of the runtime to have that symlink automatically.

**Comment by Kai-AMD** [comment]
Hi @smcv,

We did following two experiments:
(1) Create a symbolic link SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd, actually the /run/host/etc/amd doesn't exist in the host system, so the dest of the link should be invalid.
(2) Create a symbolic link SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /etc/amd.

But both of them can't solve the problem.

Thanks,
Kai

**Comment by smcv** [comment]
> actually the /run/host/etc/amd doesn't exist in the host system

I know, but it does exist inside the container.

**Comment by smcv** [comment]
Please try running Steam with `PRESSURE_VESSEL_SHELL=instead` in the environment, then try to launch a game in the container. When you do this, instead of the actual game running, you will get an `xterm` with a shell inside the container, which you can use to explore the container interactively and figure out whether anything else the driver needs is missing. I think the best way to get this working will probably be for someone who understands the AMDVLK driver's requirements to explore this environment and check for missing files.

If you have a symbolic link `SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd` on the host system, then you should find that inside the container, `/etc/amd` is a symbolic link to `/run/host/etc/amd`, which has the content from `/etc/amd` on the host system. This is because the host system's `/etc` and `/usr` are mounted on `/run/host/etc` and `/run/host/usr` inside the container, similar to what happens in Flatpak.

**Comment by smcv** [comment]
> Create a symbolic link SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /etc/amd

This is never going to be helpful, because `files/etc` gets mounted on `/etc` inside the container, so the result would be a recursive symlink `/etc/amd -> /etc/amd`.

**Comment by WenqingLiAMD** [comment]
> Please try running Steam with `PRESSURE_VESSEL_SHELL=instead` in the environment, then try to launch a game in the container. When you do this, instead of the actual game running, you will get an `xterm` with a shell inside the container, which you can use to explore the container interactively and figure out whether anything else the driver needs is missing. I think the best way to get this working will probably be for someone who understands the AMDVLK driver's requirements to explore this environment and check for missing files.
> 
> If you have a symbolic link `SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd` on the host system, then you should find that inside the container, `/etc/amd` is a symbolic link to `/run/host/etc/amd`, which has the content from `/etc/amd` on the host system. This is because the host system's `/etc` and `/usr` are mounted on `/run/host/etc` and `/run/host/usr` inside the container, similar to what happens in Flatpak.

Hi @smcv,

I create the link `SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd`, set the environment `PRESSURE_VESSEL_SHELL=instead`, then start the xterm, but still no `/run/host/etc/amd` found, please see this screenshot:

![steam_runtime_path](https://user-images.githubusercontent.com/65225636/105660542-2fb89300-5f06-11eb-9a4b-29ec14030588.png)

Here is a rough summary of the paths driver needs at runtime:
1. settings: `/etc/amd/`
2. config: `/usr/share/vulkan/`
3. cache: `$HOME/.cache/`
4. log&debug: `/tmp/`, `/var/tmp/`

**Comment by smcv** [comment]
> I create the link SteamLinuxRuntime_soldier/soldier/files/etc/amd -> /run/host/etc/amd, set the environment PRESSURE_VESSEL_SHELL=instead, then start the xterm, but still no /run/host/etc/amd found

Thanks, I'll look into that. We might need to alter the C code: I thought we already mounted all subdirectories of the host `/etc` in the container's `/run/host/etc`, but from your testing it seems we don't. We probably should.

> vulkan json file

We should already be looking in all the paths used in the official [Vulkan-Loader](https://github.com/KhronosGroup/Vulkan-Loader) ([`get_vulkan_search_paths()` in `steam-runtime-tools/graphics.c`](https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/blob/master/steam-runtime-tools/graphics.c), called with `suffix = "vulkan/icd.d"`).

This includes all the ICD paths you mentioned, except for `/usr/local/etc/vulkan/icd.d`, which would not be used unless you have a self-compiled Vulkan-Loader built with `CMAKE_INSTALL_PREFIX=/usr/local`. We do not currently have a good way to introspect what the Vulkan-Loader's `CMAKE_INSTALL_PREFIX` was, so we assume `/usr`, which should be right for most (all?) Linux distros.

We currently put the JSON files (possibly with their contents rewritten to point to appropriate paths inside the container) in `/overrides/share/vulkan/icd.d` inside the container, but we set `XDG_DATA_DIRS` to make sure that path gets searched by the Vulkan-Loader inside the container.

> amd cache: `$HOME/.cache/AMD`

At the moment we share the whole home directory with the container. However, future work might give each game its own private home directory. If that happens, then each game will have its own private `$HOME/.cache/AMD`.

> log&debug: `/tmp/`, `/var/tmp/`

pressure-vessel currently shares the host's `/tmp` with each game, but this is not guaranteed: games might get their own private `/tmp` in future.

I'm honestly not sure what we do with `/var/tmp` right now. It might either be private to the game, or sharing the host `/var/tmp`.

**Comment by smcv** [comment]
> We might need to alter the C code

Yes we do. https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/232

> I'm honestly not sure what we do with /var/tmp right now

It turns out it didn't exist at all, which was probably not very helpful either.

If you're comfortable with trying unreleased/unreviewed code, perhaps you could try replacing the `SteamLinuxRuntime_soldier/pressure-vessel/` directory with the result of unpacking <https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/jobs/27154/artifacts/file/_build/scout/pressure-vessel-0.20210126.0.10-53d4-bin+src.tar.gz>?

**Comment by WenqingLiAMD** [comment]
Yes, amd setting files under `/etc/amd` works! I can also see the `/run/host/etc` in the xterm is identical to the `/etc` on host.

**Comment by smcv** [comment]
Thanks, that change will be in the next beta (I don't know when that will be, but it will be versioned 0.20210128 or later).

**Comment by smcv** [comment]
Steam Runtime 'soldier' builds >= 0.20210201.0 should have this fixed. Please re-test?

**Comment by WenqingLiAMD** [comment]
> Steam Runtime 'soldier' builds >= 0.20210201.0 should have this fixed. Please re-test?

All paths are available now when testing with builds 0.20210208.0, many thanks.

**Comment by smcv** [comment]
@kisak-valve or @Kai-AMD, please could you close this? Confirmed fixed by the previous comment from @WenqingLiAMD.

**Comment by kisak-valve** [comment]
Closing as fixed in the soldier	0.20210208.0 build which has shipped on the mainline branch of Steam Linux Runtime - Soldier.
